tools/autograd/templates/VariableType.cpp

```diff
@@ -173,12 +173,61 @@ Variable VariableType::as_variable(const Scalar & scalar) const {
   return make_variable(std::move(tensor));
 }
 
-static void check_inplace(const VariableImpl& pImpl) {
-  if (pImpl.requires_grad && !pImpl.grad_fn) {
+struct VariableFlags {
+  bool requires_grad;
+  bool is_volatile;
+};
+
+template<typename T>
+static VariableFlags compute_flags_tmpl(T tensors) {
+  VariableFlags flags = {false, false};
+  for (const Tensor& tensor : tensors) {
+    auto& var = static_cast<const Variable&>(tensor);
+    if (var.defined()) {
+      flags.requires_grad |= var.requires_grad();
+      flags.is_volatile |= var.is_volatile();
+    }
+  }
+  flags.requires_grad &= !flags.is_volatile;
+  return flags;
+}
+
+using TensorRef = std::reference_wrapper<const Tensor>;
```

Comment:
Is this the same thing as what I described here? https://github.com/zdevito/ATen/issues/114 (I guess not, since reference_wrapper is copy assignable)

Reply:
No. The point of this is to avoid the copy-constructor (and atomic ops) in `std::initializer_list<Tensor>`. Since `std::initializer_list<Tensor&>` isn't valid C++ (`Tensor&` isn't copyable), we have to use a `reference_wrapper`.